chore: remove KongIngress support for Gateway API Route objects and Services referenced by those Routes

pmalek opened the pull request:
**What this PR does / why we need it**:

In order to prevent users on relying on `KongIngress` annotation overrides this PR disables the ability to do so for:

- Gateway API *`Route` objects
- kubernetes `Service`s defined as Gateway API *`Route`'s backends

**Which issue this PR fixes**:

fixes #2505 

**Special notes for your reviewer**:

N/A

**PR Readiness Checklist**:

Complete these before marking the PR as `ready to review`:

- [x] the `CHANGELOG.md` release notes have been updated to reflect any significant (and particularly user-facing) changes introduced by this PR

rainest reviewed (changes requested):
Based on

```
09:42:36-0700 esenin $ cd src/gateway-api 
09:42:41-0700 esenin $ grep -r '"UDPRoute"'
pkg/client/clientset/versioned/typed/apis/v1alpha2/fake/fake_udproute.go:var udproutesKind = schema.GroupVersionKind{Group: "gateway.networking.k8s.io", Version: "v1alpha2", Kind: "UDPRoute"}
```
it does not appear there's a constant. Is that typically expected? Briefly looking through the Kubernetes core codebase, it looks like they do use the hardcoded string throughout, and that seems okay since any change to the kind name would mean that it's a different kind entirely.

The warning would be good--arguably mandatory if we're breaking the existing contract around them applying to Services always.

It is true that [FillOverrides()](https://github.com/Kong/kubernetes-ingress-controller/blob/v2.3.1/internal/dataplane/kongstate/kongstate.go#L135-L176) does not have its own information on what ingresslike kind links to a Service, and that we cannot disable KongIngress use with Gateway Route-attached Services from there as-is. To provide that information we probably need to modify the [kongstate Service type](https://github.com/Kong/kubernetes-ingress-controller/blob/v2.3.1/internal/dataplane/kongstate/service.go#L16-L23) to include the parent kind, which we'd set from the various [translate functions](https://github.com/Kong/kubernetes-ingress-controller/blob/v2.3.1/internal/dataplane/parser/translate_udproute.go#L169-L187). 

From there we could add cases to `continue` in the `FillOverrides()` service and upstream loops.

pmalek commented:
Alright so I've added some tests for annotations on Gateway API `*Routes` and added a warning when that happens.

As for the override annotations on Services referenced by `*Routes`: it seems that we now have (unreleased yet) support for multiple backend refs in `*Routes` [which is stored/treated(?) on the same level as `kong.Service`](https://github.com/Kong/kubernetes-ingress-controller/blob/35de6ee205a7e7c64f240aab8c471f947d6f33dd/internal/dataplane/parser/translate_utils.go#L169-L179).
So my understanding is that there is a 1 to n mapping between a `kong.Service` and `backendRef`s.
These 2 fields in turn are then stored in `kongstate.Service` which gets its overrides in [`FillOverrides()`](https://github.com/Kong/kubernetes-ingress-controller/blob/dbbe000aaf6e734df4f5cbe54b9c26e358c7ee78/internal/dataplane/kongstate/kongstate.go#L137-L146).

So the question here is:
how does one map the overrides that can happen for any `Service` referenced by a `*Route` object (via `backendRefs`), where the [overrides are applied on the `kongstate.Service`](https://github.com/Kong/kubernetes-ingress-controller/blob/03fe8c9bd6b313eefd2e46382065890e24a626b4/internal/dataplane/kongstate/service.go#L116-L129) (if I read it right this maps 1-1 to `kong.Service`, which is what ends up in Admin API)?

pmalek commented:
> we'll also want to change
> 
> https://github.com/Kong/kubernetes-ingress-controller/blob/52c0e00d04610d9c7b64a233a4e615ca80b52c01/internal/dataplane/parser/ingressrules.go#L58-L61
> 
> to refer to annotations in general--it should have said that already, since it always applied to any annotation, not just override and the attached KongIngress settings.

What specifically would you like to see mentioned here? What other objects can be set as backends to mention them there?

rainest commented:
> So my understanding is that there is a 1 to n mapping between a kong.Service and backendRefs
> [With an HTTPRoute that has two Services in its BackendRef] I get just 1 Service, 1 Upstream and 1 Route in Admin API. What am I missing here?

We had to fit multiple backendRefs into our model a bit differently than you might immediately expect based on the names. Kong routes can point to one Kong service and one Kong service only, so we have to fit multiple Kubernetes Services into a single Kong service. We use our services more to indicate connection and request characteristics (connect timeout, upstream protocol, upstream hostname, etc.), and then the upstream endpoint, but that endpoint can take several forms.

A Kong service host can indicate either a regular, resolves with `dig example.com` DNS hostname or a virtual DNS hostname that only Kong can resolve internally. These virtual entries are Kong upstreams, and our DNS "resolution" for them attempts to map `service.host` -> `upstream.name`. If there is an upstream whose name matches the service host, Kong's DNS resolver (the ["balancer"](https://github.com/Kong/kong/tree/2.8.1/kong/runloop/balancer)) will return the upstream's targets instead of the DNS A records you'd get from actual DNS resolution. The balancer is also responsible for load balancing (`upstream.algorithm`, `upstream.hash_on`, and `target.weight`) and healthchecking (`upstream.healthchecks`), so within Kong it only returns a single address to the router, a healthy address chosen based on the LB algorithm, so we set LB/target weights to implement `BackendRef.weight`.

Kong's model has the upstream's targets hold the actual individual endpoints/addresses, so to satisfy the HTTPRoute, we create the single Kong service to hold service settings and an upstream whose targets are the collection of endpoints from each Kubernetes service (or each Kubernetes service hostname with service-upstream). If you wanted to see the multiple Kubernetes Services represented, you'd want to look at https://docs.konghq.com/gateway/2.8.x/admin-api/#list-targets in the admin API. The KongIngress (and annotation) overrides apply to the single Kong service and single Kong upstream we create for the route (targets have no tunables other than the per-target weight). There's no way for us to handle multiple Kubernetes Services within a BackendRef having different overrides because we create only one Kong upstream and service for the entire BackendRef, so we enforce all Services having the same overrides.

rainest commented:
> What specifically would you like to see mentioned here? What other objects can be set as backends to mention them there?

Specifically I was looking at removing the now irrelevant KongIngress mention and indicating that it's any of our annotations. May as well try and simplify the prose in general and change the info about where this matters:

```
return fmt.Errorf("the Kubernetes Services %v cannot have different sets of konghq.com annotations" +
"These Services are used in the same Gateway Route BackendRef together to create the Kong Service %s" +
"and must use the same Kong annotations", k8sServices, *service.Name)
```

This condition actually only affects Gateway Routes (not an official term, looking for consensus in [SIG chat](https://kubernetes.slack.com/archives/CR0H13KGA/p1654879571346699)); we wrote it for a potential future where we may have other types of config that can use multiple Services. And now, scope creep!

IMO that language may be a bit confusing if you're looking at the error and trying to figure out what generated it, and while my version narrows the scope, it doesn't really help you find the precise object--the user will want to look at the specific broken HTTPRoute or whatever, not scan through all Gateway Routes. This is separate since it's not truly needed--we already provide the Services, and that's all you need to fix, but it's probably useful to know what's using them rather than changing them blindly--you won't necessarily know what annotations are appropriate to add or remove without knowing what they're being used from: https://github.com/Kong/kubernetes-ingress-controller/issues/2566

To indicate that we probably actually want to populate an additional `ingressRules.ServiceNameToParent` map that points to a GVK and NamespacedName so that we can:

```
return fmt.Errorf("the Kubernetes Services %v cannot have different sets of konghq.com annotations" +
"These Services are used in %s together to create the Kong Service %s" +
"and must use the same Kong annotations",
k8sServices, <some string representation of the ingressRules.ServiceNameToParent>, *service.Name)
```

_Finally_ the way we handle this error is maybe not great. Prior to adding that error, `populateServices()` always succeeded. If it now fails, that failure [bubbles up to the parser build and indicates that it failed completely](https://github.com/Kong/kubernetes-ingress-controller/blob/52c0e00d04610d9c7b64a233a4e615ca80b52c01/internal/dataplane/parser/parser.go#L79-L81), and the parser build bubbles that up saying that it [also failed completely](https://github.com/Kong/kubernetes-ingress-controller/blob/52c0e00d04610d9c7b64a233a4e615ca80b52c01/internal/dataplane/kong_client.go#L307-L313), invalidating the entire config. That's a bit aggressive for something that only really breaks one Gateway Route--we could proceed with the others fine despite one having invalid Service configuration, we'd just need to skip the broken one (and probably lack the ability to do that easily).

The actual end state we probably want to reach is that those errors just invalidate the one Gateway Route and log as much (https://github.com/Kong/kubernetes-ingress-controller/issues/2565). Ideally we'd be able to use a Condition on the object itself, but [there isn't an appropriate one yet](https://github.com/kubernetes-sigs/gateway-api/blob/9186abe0a9f9267a04ab86332d440bab91775a9d/apis/v1alpha2/shared_types.go#L219-L243) (we may want to consider requesting one from upstream). Failing piecemeal would then let us just log each, and bonus points use structured logs correctly:

```
log.WithFields(logrus.Fields{
"parent_gvk": <the parent GVK>,
"parent_namespace": <the parent namespace>,
"parent_name": <the parent name>,
"kong_service": *service.Name,
"kubernetes_service": <CSV of Service namespaced names>, // is there a good way to handle variable arity fields with logrus? it's a map key, so you can't duplicate :(
}).Errorf("services used within a single backend have different konghq.com annotations")

pmalek commented:
Thanks for those extensive comments and reporting those other issues. That helps a lot!

Now in order to minimize the scope of this PR - since we already have those other issue that you've reported and which are pretty specific - what do you suggest we cover in this PR from the issues that you've enlisted?

I reckon we need to set the Condition on the - if I understood correctly - Route object? Yet you mentioned as well  that 

> Note that the parser cannot set resource status information directly: only the controller reconcilers can do so. To communicate information back from the parser to the controllers, other controllers use a [StatusQueue](https://github.com/Kong/kubernetes-ingress-controller/blob/ec8bde7aed490e4f1dc7bad4f3742a8edc58e57e/internal/controllers/configuration/zz_generated_controllers.go#L302-L313) watch that [reads address status information](https://github.com/Kong/kubernetes-ingress-controller/blob/ec8bde7aed490e4f1dc7bad4f3742a8edc58e57e/internal/controllers/configuration/zz_generated_controllers.go#L404-L424). As best I can tell, the ones we use for Ingress and such read [static information](https://github.com/Kong/kubernetes-ingress-controller/blob/ec8bde7aed490e4f1dc7bad4f3742a8edc58e57e/internal/manager/setup.go#L221-L247), and the parser queue event only triggers that read (@shaneutt can explain this system better), so a Condition setter that'd be able to handle dynamic info from the parser would be fairly different. **Since we don't have a drop-in solution, it may make sense to defer that til we can handle it for more than one Route at a time.**

In which case: what would you suggest to cover in this PR :)? I believe adding the parent obj to the log via an addition of `ingressRules.ServiceNameToParent` should be relatively easy (as this PR was meant to be 🙃 )

rainest commented:
For this PR, I'd recommend just changing the message to be about annotations instead of KongIngress, so the

```
return fmt.Errorf("the Kubernetes Services %v cannot have different sets of konghq.com annotations" +
"These Services are used in the same Gateway Route BackendRef together to create the Kong Service %s" +
"and must use the same Kong annotations", k8sServices, *service.Name)
```

The conditions would be good and more aligned with the wants of the Gateway spec, but given the need to build out a new system to shuttle them to the reconciler and the limitation around marking only one route, more work than I think we want here for limited benefit. I've edited #2566 to mention that, since it's what will enable setting conditions for all problem Routes, not just the first one the parser encounters.

pmalek commented:
Thanks! Changed the error msg. PTAL.

rainest commented on internal/dataplane/kongstate/upstream.go:
```diff
@@ -79,12 +82,45 @@ func (u *Upstream) overrideByKongIngress(kongIngress *configurationv1.KongIngres
+		// skip it since those should not be affected.
+		gvk := u.Service.Parent.GetObjectKind().GroupVersionKind()
+		if gvk.Group == gatewayv1alpha2.GroupName {
+			switch gvk.Kind {
+			case "HTTPRoute", "UDPRoute", "TCPRoute", "TLSRoute":
```
Ditto switch comment from earlier.

pmalek commented on internal/dataplane/kongstate/route.go:
```diff
@@ -235,6 +236,22 @@ func (r *Route) override(log logrus.FieldLogger, kongIngress *configurationv1.Ko
+			}).Warn("KongIngress annotation is non-actionable on Gateway API *Route objects.")
+			return
+		}
+	}
```
Agreed and removed 👍
